client/allocrunner/alloc_runner_test.go

```diff
@@ -1568,3 +1568,220 @@ func TestAllocRunner_PersistState_Destroyed(t *testing.T) {
 	require.NoError(t, err)
 	require.Nil(t, ts)
 }
+
+func TestAllocRunner_Reconnect(t *testing.T) {
+	t.Parallel()
+
+	type tcase struct {
+		clientStatus string
+	}
+	tcases := []tcase{
+		{
+			structs.AllocClientStatusRunning,
+		},
+		{
+			structs.AllocClientStatusComplete,
+		},
+		{
+			structs.AllocClientStatusFailed,
+		},
+		{
+			structs.AllocClientStatusPending,
+		},
+	}
+
+	for _, tc := range tcases {
+		t.Run(tc.clientStatus, func(t *testing.T) {
+			// create a running alloc
+			current := mock.BatchAlloc()
+
+			// Ensure task takes some time
+			task := current.Job.TaskGroups[0].Tasks[0]
+			task.Driver = "mock_driver"
+			task.Config["run_for"] = "30s"
+
+			conf, cleanup := testAllocRunnerConfig(t, current)
+			defer cleanup()
+
+			ar, err := NewAllocRunner(conf)
+			require.NoError(t, err)
+			defer destroy(ar)
+
+			go ar.Run()
+
+			updater := conf.StateUpdater.(*MockStateUpdater)
+			var last *structs.Allocation
+			testutil.WaitForResult(func() (result bool, err error) {
+				result = false
+				last = updater.Last()
+				if last == nil {
+					return
+				}
+
+				if last.ClientStatus != structs.AllocClientStatusRunning {
+					return
+				}
+
+				result = true
+				return
+			}, func(err error) {
+				require.NoError(t, err)
```

Comment:
[`WaitForResult`](https://github.com/hashicorp/nomad/blob/v1.2.6/testutil/wait.go#L17-L35) unfortunately does not have a similar API to testify's `Eventually`. Your test function doesn't return an error so this will never fail even if you run out of retries. (This comment applies to all the uses of `WaitForResult` in this PR.)